amplicon_stats.c

```diff
@@ -0,0 +1,924 @@
+                printf("\t%.1f", stats->covered_perc[i][d] / nfile);
+            }
+            printf("\n");
+
+            printf("CPCOV-%d\tSTDDEV", args->min_depth[d]);
+            for (i = 0; i < namp; i++) {
+                double n1 = stats->covered_perc[i][d] / nfile;
+                double v = stats->covered_perc2[i][d] / nfile - n1*n1;
+                printf("\t%.1f", sqrt(v));
+            }
+            printf("\n");
+        } while (++d < MAX_DEPTH && args->min_depth[d]);
+    }
+
+
+    // TLEN distribution
+    printf("# Distribution of aligned template length.\n");
+    printf("# Use 'grep ^%cTLEN | cut -f 2-' to extract this part.\n", type);
+    printf("%cTLEN\t%s\t0", type, name); // all merged
+    int tlen_dist[MAX_TLEN] = {0};
+    for (i = 0; i < namp; i++) // accumulate for all amps
+        for (j = 0; j < MAX_TLEN; j++)
+            tlen_dist[j] += stats->tlen_dist[i][j];
+    for (j = 0; j < MAX_TLEN; j++)
+        printf("\t%d", tlen_dist[j]);
+    printf("\n");
+
+    for (i = 0; i < namp; i++) {
+        printf("%cTLEN\t%s\t%d", type, name, i+1); // per amplicon
+        for (j = 0; j < MAX_TLEN; j++)
+            printf("\t%d", stats->tlen_dist[i][j]);
+        printf("\n");
+    }
+
+
+    // AMP length distribution.
+    // 0 = both ends in this amplicon
+    // 1 = ends in different amplicons
+    // 2 = other end matching an unknown amplicon site
+    //     (see tlen_dist for further analysis of where)
+    printf("# Classification of amplicon status.  Columns are\n");
+    printf("# number with both primers from this amplicon, number with\n");
+    printf("# primers from different amplicon, and number with a position\n");
+    printf("# not matching any valid amplicon primer site\n");
+    printf("# Use 'grep ^%cAMP | cut -f 2-' to extract this part.\n", type);
+    printf("%cAMP\t%s\t0", type, name); // all merged
+    int amp_dist[3] = {0};
+    for (i = 0; i < namp; i++) { // accumulate for all amps
+        amp_dist[0] += stats->amp_dist[i][0];
+        amp_dist[1] += stats->amp_dist[i][1];
+        amp_dist[2] += stats->amp_dist[i][2];
+    }
+    printf("\t%d\t%d\t%d\n", amp_dist[0], amp_dist[1], amp_dist[2]);
+
+    for (i = 0; i < namp; i++) {
+        printf("%cAMP\t%s\t%d", type, name, i+1); // per amplicon
+        printf("\t%d\t%d\t%d\n", stats->amp_dist[i][0],
+               stats->amp_dist[i][1], stats->amp_dist[i][2]);
+    }
+
+//    for (i = 0; i < namp; i++) {
+//        printf("%cAMP\t%s\t%d", type, name, i+1); // per amplicon
+//        for (j = 0; j < namp; j++)
+//            printf("\t%d", stats->amp_pair[i][j]);
+//        printf("\n");
+//    }
+
+}
+
+char const *get_sample_name(sam_hdr_t *header, char *RG) {
+    kstring_t ks = {0};
+    sam_hdr_find_tag_id(header, "RG", RG?"ID":NULL, RG, "SM", &ks);
+    return ks.s;
+}
+
+int64_t get_ref_len(sam_hdr_t *header, char *SQ) {
+    kstring_t ks = {0};
+    sam_hdr_find_tag_id(header, "SQ", SQ?"ID":NULL, SQ, "LN", &ks);
```

Comment:
If this fails, `ks.s` will be NULL, causing a blow-up on the next line.
You could also instead try:
```
int tid = SQ ? sam_hdr_name2tid(header, SQ) : 0;
return tid >= 0 ? sam_hdr_tid2len(header, tid) : -1;
```
which removes the need for the kstring.